### PR TITLE
dialyzer: fix singleton type var in union types

### PR DESCRIPTION
**Pull request by kikofernandez**
Dialyzer (or `erl_lint.erl`) was accepting code of the following form:

```erlang
-spec run_test(Opts) -> term()
      when Opts :: {join_specs, Bool} | {test, Bool}.
run_test(Opts) ->    {error, fail}.
```

But this cannot be ever checked by Dialyzer as the type variable `Bool` is in an union, which means that `Bool` happened only once and thus, `Bool` is an unbound type variable. Essentially, the example above should be rejected by Dialyzer as is the one below:

```erlang
-spec run_test_error(Opts) -> term()
      when Opts :: {join_specs, Bool}.
run_test_error(Opts) ->
    {error, fail}.
```

This commit fixes GH-6508 and adds a bunch of tests to `erl_lint.erl` so that we can keep track of which singleton type variables should throw an error and which ones are ok.

### CODE REVIEW

File: lib/kernel/src/wrap_log_reader.erl [hunk 1]
```diff
@@ -116,11 +116,11 @@ open(File, FileNo) when is_list(File), is_integer(FileNo) ->
 close(#wrap_reader{fd = FD}) ->
     file:close(FD).
 
--type chunk_ret() :: {Continuation2, Terms :: [term()]}
-                   | {Continuation2,
+-type chunk_ret() :: {Continuation2 :: term(), Terms :: [term()]}
```
Comment (kikofernandez) [inline lib/kernel/src/wrap_log_reader.erl:119]: Errors detected by the new linter / dialyzer

**Review by jhogberg — APPROVED** [review]

**Comment by essen** [comment]
Please note this a potential incompatibility in the release notes because this breaks compilation of code that was compiling for many years.

**Comment by kikofernandez** [comment]
It will be marked as a potential incompatibility in the release notes

**Comment by josevalim** [comment]
@kikofernandez, would it be possible to ship this as a warning, at least for Erlang/OTP 26?

This is restricting the ability of projects to try out Erlang/OTP 26, because working code no longer compiles (and there were no past warnings), and I assume it may also restrict the ability of said projects to migrate to Erlang/OTP 26 once it is out. If this could be a warning, it would at least give folks a year to update their project and dependencies.

Thank you.

**Comment by kikofernandez** [comment]
We will try to make it a warning now, and an error in OTP-27

**Pull request merged**